### PR TITLE
multiple fixes for java mirrors  

### PR DESCRIPTION
**Pull request by xeno-by**
First of all in 911bbc4fdd I've completely overlooked the fact that
reflectConstructor exists and that is also needs sanity checks.

Now reflectConstructor checks that the incoming symbol is actually a ctor,
and that it is actually a ctor of the class reflected by the current mirror.

Secondly while fixing reflectConstructor I've found out that reflectXXX methods
don't accept inherited, but overriden members. That's also fixed now.

Haha that was a nice one. After a fixed the problem with inherited members,
I've uncovered another issue. If we have a field overriden in a descendant,
then (unlike with methods) overriding and overriden members make sense
independently of each other. Hence in `FieldMirror.get` we need some way
of specifying whether we want the base value or the overriding value.
To do that I introduced an optional parameter. Luckily we don't need to change
`FieldMirror.set` because vars cannot be overriden.

### CODE REVIEW

**Comment by xeno-by** [comment]
The pull request depends on https://github.com/scala/scala/pull/1048, because both of them touch `checkMemberOf`, so separating them would lead to merge conflicts

**Comment by xeno-by** [comment]
review by @adriaanm or @paulp

**Comment by paulp** [comment]
It is likely we will be able to override vars in the future.  (You can do it now with -Yoverride-vars.) It would not be good to bake it into the API that we can't.

**Comment by xeno-by** [comment]
@paulp updated the pull request

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 1]
```diff
@@ -279,7 +332,11 @@ trait JavaMirrors extends internal.SymbolTable with api.JavaUniverse { self: Sym
             extends JavaTemplateMirror with ClassMirror {
       def erasure = symbol
       def isStatic = false
-      def reflectConstructor(constructor: MethodSymbol) = new JavaConstructorMirror(outer, constructor)
+      def reflectConstructor(constructor: MethodSymbol) = {
+        if (!constructor.isClassConstructor) ErrorNotConstructor(constructor, symbol)
+        if (!symbol.info.decls.toList.filter(_.isConstructor).contains(constructor)) ErrorNotConstructor(constructor, symbol)
```
Comment (adriaanm) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: why do you need the filter?

**Comment by adriaanm** [comment]
overlapping with #1052, btw

**Comment by xeno-by** [comment]
That's intended, as described in comments to #1052 and #1054

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: true that, I don't. fixed

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: src/reflect/scala/reflect/api/Mirrors.scala [hunk 2]
```diff
@@ -104,21 +104,40 @@ trait Mirrors { self: Universe =>
...
+     *  The optional `preferOverriding` parameter governs that choice
+     *  (if the field is not overriden, then the value of this parameter is ignored).
+     *  Default behaviour is to prefer the overriding value.
+     *
      *  Scala reflection uses reflection capabilities of the underlying platform,
      *  so `FieldMirror.get` might throw platform-specific exceptions associated
      *  with getting a field or invoking a getter method of the field.
      */
-    def get: Any
+    def get(preferOverriding: Boolean = true): Any
```
Comment (paulp) [inline src/reflect/scala/reflect/api/Mirrors.scala:119]: For what little it is worth, my opinion has not changed in the least that Boolean parameters in public API are awful and Booleans with defaults are worse.

```
 get(true)
```

WTF is that? Why kid ourselves that everyone everywhere will write "get(preferOverriding = true)" when it's so obviously not true?

Comment (adriaanm) [inline src/reflect/scala/reflect/api/Mirrors.scala:119]: I agree. Maybe two littles make enough of worth?

**Comment by odersky** [comment]
I am against having special logic in Field.get. Instead, we should steer people away from fields. Have a doc comment like:

"FieldMirrors are the only way to get at private[this] vals and vars and might be useful to inspect the data of underlying Java fields. For all other uses, it's better to go through the fields accessor. In particular, there should be no need to ever access a field mirror when reflecting on just the public members of a class or trait.

Note also that only accessor MethodMirrors, but not FieldMirrors will accurately reflect overriding behavior."

Another note:

isClassConstructor/isMixinConstructor should be combined into one method: isConstructor. You can tell which one it is by looking atthe parent.

**Comment by xeno-by** [comment]
Will be resubmitted jointly with https://github.com/scala/scala/pull/1048, https://github.com/scala/scala/pull/1052, https://github.com/scala/scala/pull/1054 and https://github.com/scala/scala/pull/1057.

**Comment by xeno-by** [comment]
Superceded by https://github.com/scala/scala/pull/1067

**Pull request closed without merging**